cpp/doc/DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,739 @@
+changed code before committing it. The easiest way to do this is to configure your editor to
+"format on save", or to use `pre-commit`.
+
+Aspects of code style not discussed in this document and not automatically enforceable are typically
+caught during code review, or not enforced.
+
+### C++ Guidelines
+
+In general, we recommend following
+[C++ Core Guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines). We also
+recommend watching Sean Parent's [C++ Seasoning talk](https://www.youtube.com/watch?v=W2tWOdzgXHA),
+and we try to follow his rules: "No raw loops. No raw pointers. No raw synchronization primitives."
+
+ * Prefer algorithms from STL and Thrust to raw loops.
+ * Prefer libcudf and RMM [owning data structures and views](#libcudf-data-structures) to raw
+   pointers and raw memory allocation.
+
+Documentation is discussed in the [Documentation Guide](DOCUMENTATION.md).
+
+### Includes
+
+The following guidelines apply to organizing `#include` lines.
+
+ * Group includes by library (e.g. cuSpatial, RMM, Thrust, STL). `clang-format` will respect the
+   groupings and sort the individual includes within a group lexicographically.
+ * Separate groups by a blank line.
+ * Order the groups from "nearest" to "farthest". In other words, local includes, then includes
+   from other RAPIDS libraries, then includes from related libraries, like `<thrust/...>`, then
+   includes from dependencies installed with cuDF, and then standard headers (for example `<string>`,
+   `<iostream>`).
+ * Use `<>` instead of `""` unless the header is in the same directory as the source file.
+ * Tools like `clangd` often auto-insert includes when they can, but they usually get the grouping
+   and brackets wrong.
+ * Always check that includes are only necessary for the file in which they are included.
+   Try to avoid excessive including especially in header files. Double check this when you remove
+   code.
+ * Use quotes `"` to include local headers from the same relative source directory. This should only
+   occur in source files and non-public header files. Otherwise use angle brackets `<>` around
+   included header filenames.
+ * Avoid relative paths with `..` when possible. Paths with `..` are necessary when including
+   (internal) headers from source paths not in the same directory as the including file,
+   because source paths are not passed with `-I`.
+ * Avoid including library internal headers from non-internal files. For example, try not to include
+   headers from libcuspatial `src` directories in tests or in libcuspatial public headers. If you
+   find yourself doing this, start a discussion about moving (parts of) the included internal header
+   to a public header.
+
+# libcuspatial Data Structures
+
+The header-only libcuspatial API is agnostic to the type of containers used by the application to
+hold its data, because the header-only API is based on iterators. The cuDF-based cuSpatial API, on 
```

Comment:
Perhaps we can talk a bit about the iterator requirements here? Like random accessible and device accessible, conforming to LAI etc.

Reply:
Done, please have a look.